lib/elixir/lib/kernel.ex

```diff
@@ -957,7 +971,10 @@ defmodule Kernel do
       [1, 2 | 3]
 
   """
-  @spec (list ++ term) :: maybe_improper_list
+  @spec (nonempty_list ++ nonempty_list) :: nonempty_list
+  @spec (left ++ []) :: left when left: nonempty_list
+  @spec ([] ++ right) :: right when right: term
+  @spec (list ++ nonempty_improper_list(any, any)) :: nonempty_improper_list(any(), any())
```

Comment:
@josevalim I realized I could simplify the specs from 4 to 2 clauses,

```elixir
  @spec ([] ++ right) :: right when right: var
  @spec (nonempty_list ++ maybe_improper_list(any, tail))
        :: nonempty_maybe_improper_list(any, tail) when tail: var
```

still the original one is simpler, but i like this one because it tells the user that if given an empty list on the left it will return whatever term on the right, making it clear it will not always return a list or an improper one as the docs currently express.

Once this PR is merged I will submit a PR improving the docs of many functions including this one.

so your call about this one and I think we are good to go.